{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# This program is designed to take as input two files:  a FASTA file of sequences, and Query file. \n",
    "# From the FASTA file, contigs are assembled, and the query file contains a sequence subsection that is\n",
    "# used to query the assembled contigs, wherein the goal is to find the longest contig containing the query sequence. \n",
    "# The outputs are as follows:\n",
    "# ALLELES.FASTA which is a FASTA file of the largest constructed contig containing the query sequence\n",
    "# ALLELES.ALN which is a tab-delimited file describing the alignment of sequence reads to the assembled contig(s) in\n",
    "# the ALLELES.FASTA file. \n",
    "# The columns in ALLELES.ALN are as follows: \n",
    "# SSEQID name of sequencing read (from READS.fastq.gz)\n",
    "# QSEQID name of contig matched (from ALLELES.fasta)\n",
    "# SSTART starting coordinate in sequencing read sseqid that matches qseq\n",
    "# SEND ending coordinate in sequencing read  sseqid that matches qseq\n",
    "# QSTART starting coordinate in contig that matches sseq\n",
    "# QEND ending coordinate in contig that matches sseq"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "source": [
    "import re \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import os.path\n",
    "# import unittest\n",
    "# from unittest import mock\n",
    "# from unittest import TestCase\n",
    "# from io import StringIO\n",
    "\n",
    "# These global variables define the path in which this workflow will be run, as well as the files for both reads and query inputs. \n",
    "# These inputs default to the local directory in which this notebook is stored, along with default filenames. \n",
    "# The Query file is expected to have a first line identifiying the file as query, if this in untrue remove the \"[1:]\" after readlines\"\n",
    "# on line 29.\n",
    "ReadDirectory = (input(\"Enter the directory of the FASTA reads file: \")) or (os.path.join(sys.path[0]))\n",
    "QueryDirectory = (input(\"Enter the directory of the QUERY query file: \")) or (os.path.join(sys.path[0]))\n",
    "ReadFilename = (input(\"Enter the filename of the FASTA reads file: \")) or (\"READS.fasta\")\n",
    "QueryFilename = (input(\"Enter the filename of the  reads file: \")) or (\"QUERY.fasta\") \n",
    "\n",
    "\n",
    "#This function takes the user inputs (or default values) and stores the information read from the files \n",
    "# Secondarily for some reason the QUERY file cannot be stripped of it's various newline and brackets when read \n",
    "def Test_FASTAreader(ReadDirectory, QueryDirectory, ReadFilename, QueryFilename):\n",
    "    global FASTAreads\n",
    "    global FASTAquery\n",
    "    ReadFile = open(os.path.join(ReadDirectory, ReadFilename))\n",
    "    QueryFile = open(os.path.join(QueryDirectory, QueryFilename))\n",
    "    FASTAreads = ReadFile.read\n",
    "    ReadFile.close\n",
    "    FASTAquery = QueryFile.readlines()[1:]\n",
    "    QueryFile.close\n",
    "    FASTAquery = str(FASTAquery)\n",
    "Test_FASTAreader(ReadDirectory, QueryDirectory, ReadFilename, QueryFilename)\n",
    "print(FASTAreads)\n",
    "print(FASTAquery)\n",
    "\n",
    "#This function takes the reads from a FASTA file and breaks those down into K-mers, and stores them as a variable\n",
    "#  for the next functions input. \n",
    "def Test_Kmers_Creation(FASTAreads):\n",
    "    pass\n",
    "#Create some variable in which K-mers are stored\n",
    "    global Kmers\n",
    "    Kmers = \"Ooof\"\n",
    "Test_Kmers_Creation(FASTAreads)\n",
    "\n",
    "# This function takes the generated Kmers from the previous function and constructs a De-bruijn graph as an adjacency matrix for \n",
    "# use by the next Eulerian path algorithm in function form. \n",
    "def Test_DeBruijn(Kmers):\n",
    "    pass \n",
    "    #Creates a De-bruijn graph constructed of the K-mers from the previous function, store as adjacency list\n",
    "    global DeBruijn_Adjacency \n",
    "    DeBruijn_Adjacency = [ 'O', 'o', 'o', 'f']\n",
    "Test_DeBruijn(Kmers)\n",
    "\n",
    "#This function uses the Debrjuin graph to find a path wherein all the edges are crossed exactly once to assemble contigs from the (K-1)mers, \n",
    "# which act as nodes, and paths are length K. \n",
    "def Test_Eulerian_Path(DeBruijn_Adjacency):\n",
    "    pass\n",
    "    #Creates the contigs from using an Eulerian path through the De-Bruijn graph, saved as Contigs variable\n",
    "    global Contigs\n",
    "    Contigs = [\"Oo\", \"of\"]\n",
    "Test_Eulerian_Path(DeBruijn_Adjacency)\n",
    "\n",
    "\n",
    "\n",
    "def Test_QuerySequenceFinder(Contigs, FASTAquery):\n",
    "    QueryForward = FASTAquery\n",
    "    QueryBackward = QueryForward[::-1]\n",
    "    print(QueryForward)\n",
    "    print(QueryBackward)\n",
    "    # Currently producing some elements that are uninteded for comparison when printed.\n",
    "    # Store longest contigs in Alleles.fasta outfile, as well as in some variable.  \n",
    "    # Need to define the comparison values to store, write them in something like a pandas dataframe, write to \n",
    "    # tab-delimited file, needed for alignment file. \n",
    "    global Name_of_variables_for_AllelesFasta\n",
    "    global Name_of_Variable_for_AlleleALN\n",
    "    Name_of_variables_for_AllelesFasta = \"The FASTA part of this is that I need to code FASTA than I am\"\n",
    "    Name_of_Variable_for_AlleleALN = \"My alignment is chaotic sad.\"\n",
    "Test_QuerySequenceFinder(Contigs, FASTAquery)\n",
    "\n",
    "def Variables_to_files(Name_of_variables_for_AllelesFasta, Name_of_Variable_for_AlleleALN):\n",
    "    pass\n",
    "Variables_to_files(Name_of_variables_for_AllelesFasta, Name_of_Variable_for_AlleleALN)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "'GGGATCGGCCATTGAACAAGATGGATTGCACGCAGGTTCTCCGGCCGCTTGGGTGGAGAGGCTATTCGGCTATGACTGGGCACAACAGACAATCGGCTGCTCTGATGCCGCCGTGTTCCGGCTGTCAGCGCAGGGGCGCCCGGTTCTTTTTGTCAAGACCGACCTGTCCGGTGCCCTGAATGAACTGCAGGACGAGGCAGCGCGGCTATCGTGGCTGGCCACGACGGGCGTTCCTTGCGCAGCTGTGCTCGACGTTGTCACTGAAGCGGGAAGGGACTGGCTGCTATTGGGCGAAGTGCCGGGGCAGGATCTCCTGTCATCTCACCTTGCTCCTGCCGAGAAAGTATCCATCATGGCTGATGCAATGCGGCGGCTGCATACGCTTGATCCGGCTACCTGCCCATTCGACCACCAAGCGAAACATCGCATCGAGCGAGCACGTACTCGGATGGAAGCCGGTCTTGTCGATCAGGATGATCTGGACGAAGAGCATCAGGGGCTCGCGCCAGCCGAACTGTTCGCCAGGCTCAAGGCGCGCATGCCCGACGGCGATGATCTCGTCGTGACCCATGGCGATGCCTGCTTGCCGAATATCATGGTGGAAAATGGCCGCTTTTCTGGATTCATCGACTGTGGCCGGCTGGGTGT\\n'\n",
      "<built-in method read of _io.TextIOWrapper object at 0x1189bd990>\n",
      "['GGGATCGGCCATTGAACAAGATGGATTGCACGCAGGTTCTCCGGCCGCTTGGGTGGAGAGGCTATTCGGCTATGACTGGGCACAACAGACAATCGGCTGCTCTGATGCCGCCGTGTTCCGGCTGTCAGCGCAGGGGCGCCCGGTTCTTTTTGTCAAGACCGACCTGTCCGGTGCCCTGAATGAACTGCAGGACGAGGCAGCGCGGCTATCGTGGCTGGCCACGACGGGCGTTCCTTGCGCAGCTGTGCTCGACGTTGTCACTGAAGCGGGAAGGGACTGGCTGCTATTGGGCGAAGTGCCGGGGCAGGATCTCCTGTCATCTCACCTTGCTCCTGCCGAGAAAGTATCCATCATGGCTGATGCAATGCGGCGGCTGCATACGCTTGATCCGGCTACCTGCCCATTCGACCACCAAGCGAAACATCGCATCGAGCGAGCACGTACTCGGATGGAAGCCGGTCTTGTCGATCAGGATGATCTGGACGAAGAGCATCAGGGGCTCGCGCCAGCCGAACTGTTCGCCAGGCTCAAGGCGCGCATGCCCGACGGCGATGATCTCGTCGTGACCCATGGCGATGCCTGCTTGCCGAATATCATGGTGGAAAATGGCCGCTTTTCTGGATTCATCGACTGTGGCCGGCTGGGTGT\\n']\n",
      "['GGGATCGGCCATTGAACAAGATGGATTGCACGCAGGTTCTCCGGCCGCTTGGGTGGAGAGGCTATTCGGCTATGACTGGGCACAACAGACAATCGGCTGCTCTGATGCCGCCGTGTTCCGGCTGTCAGCGCAGGGGCGCCCGGTTCTTTTTGTCAAGACCGACCTGTCCGGTGCCCTGAATGAACTGCAGGACGAGGCAGCGCGGCTATCGTGGCTGGCCACGACGGGCGTTCCTTGCGCAGCTGTGCTCGACGTTGTCACTGAAGCGGGAAGGGACTGGCTGCTATTGGGCGAAGTGCCGGGGCAGGATCTCCTGTCATCTCACCTTGCTCCTGCCGAGAAAGTATCCATCATGGCTGATGCAATGCGGCGGCTGCATACGCTTGATCCGGCTACCTGCCCATTCGACCACCAAGCGAAACATCGCATCGAGCGAGCACGTACTCGGATGGAAGCCGGTCTTGTCGATCAGGATGATCTGGACGAAGAGCATCAGGGGCTCGCGCCAGCCGAACTGTTCGCCAGGCTCAAGGCGCGCATGCCCGACGGCGATGATCTCGTCGTGACCCATGGCGATGCCTGCTTGCCGAATATCATGGTGGAAAATGGCCGCTTTTCTGGATTCATCGACTGTGGCCGGCTGGGTGT\\n']\n",
      "]'n\\TGTGGGTCGGCCGGTGTCAGCTACTTAGGTCTTTTCGCCGGTAAAAGGTGGTACTATAAGCCGTTCGTCCGTAGCGGTACCCAGTGCTGCTCTAGTAGCGGCAGCCCGTACGCGCGGAACTCGGACCGCTTGTCAAGCCGACCGCGCTCGGGGACTACGAGAAGCAGGTCTAGTAGGACTAGCTGTTCTGGCCGAAGGTAGGCTCATGCACGAGCGAGCTACGCTACAAAGCGAACCACCAGCTTACCCGTCCATCGGCCTAGTTCGCATACGTCGGCGGCGTAACGTAGTCGGTACTACCTATGAAAGAGCCGTCCTCGTTCCACTCTACTGTCCTCTAGGACGGGGCCGTGAAGCGGGTTATCGTCGGTCAGGGAAGGGCGAAGTCACTGTTGCAGCTCGTGTCGACGCGTTCCTTGCGGGCAGCACCGGTCGGTGCTATCGGCGCGACGGAGCAGGACGTCAAGTAAGTCCCGTGGCCTGTCCAGCCAGAACTGTTTTTCTTGGCCCGCGGGGACGCGACTGTCGGCCTTGTGCCGCCGTAGTCTCGTCGGCTAACAGACAACACGGGTCAGTATCGGCTTATCGGAGAGGTGGGTTCGCCGGCCTCTTGGACGCACGTTAGGTAGAACAAGTTACCGGCTAGGG'[\n"
     ]
    }
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.7.3",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.7.3 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}